src/Parallel/Reduction.hpp

```diff
@@ -9,36 +9,87 @@
 #include "Parallel/CharmRegistration.hpp"
 #include "Parallel/ConstGlobalCache.hpp"
 #include "Parallel/PupStlCpp11.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/Functional.hpp"
+#include "Utilities/Gsl.hpp"
 #include "Utilities/Requires.hpp"
 #include "Utilities/TypeTraits.hpp"
 
 namespace Parallel {
+/// \cond
+template <class... Ts>
+struct ReductionData;
+/// \endcond
+
+namespace detail {
 /*!
  * \ingroup ParallelGroup
- * \brief Used for reducing heterogeneous collection of types in a single
- * reduction call
+ * \brief Convert a `ReductionData` to a `CkReductionMsg`. Used in custom
+ * reducers.
+ */
+template <class... Ts>
+CkReductionMsg* new_reduction_msg(
+    ReductionData<Ts...>& reduction_data) noexcept {
+  return CkReductionMsg::buildNew(static_cast<int>(reduction_data.size()),
+                                  reduction_data.packed().get());
+}
+}  // namespace detail
+
+/*!
+ * \ingroup ParallelGroup
+ * \brief The data to be reduced, and invokables to be called whenever two
+ * reduction messages are combined and after the reduction has been completed.
  *
- * For each `ReductionData` you must write a custom reducer to reduce the data.
- * Here is an example of such a function:
+ * `InvokeCombine` is a binary invokable that maps `(T current_state, T element)
+ * -> T`, where the `current_state` is the result of reductions so far. The
+ * `InvokeFinal` is an n-ary that takes as its first argument a `T
+ * result_of_reduction` and is invoked once after the reduction is completed.
+ * The additional arguments correspond to the resultant data of earlier
+ * ReductionDatum template parameters in the ReductionData, and are identified
+ * via the `InvokeFinalExtraArgsIndices`, which must be a `std::index_sequence`.
```

Comment:
Can you explain how this template parameter works in a bit more detail? Do I understand correctly that it takes a list of indices that refer to the position of other ReductionDatum parameters, and passes the result of those to the function?

I think this is a lot of complexity for the only benefit (that I can see) to not having to do the post-processing of the reduced data in the receiving action. Why not just take the sqrt and divide by that number explicitly in the action (refers to the example below)? I think that would be clearer code. Or what’s the benefit of having the reduction do that?

Reply:
I'll expand in the documentation as well, the idea is it's like indexing into a vector but you can only retrieve the elements before the one you are currently operation on. Does that make sense?

I disagree with you. There are huge benefits to doing it this way:
1. All the code necessary to understand the result of a reduction call is in one place (makes it easier to understand)
2. The action does not need to know anything about how the reduction was done (important for observers where I may want the L2 norm of one quantity and L1 of another. Here I can just change the finalize invokable and I've made the change, no need to change the action)

Reply:
sounds good